src/cursor.rs

```diff
@@ -456,25 +453,37 @@ impl SyntaxNode {
             return TokenAtOffset::None;
         }
 
-        let mut children = self.children_with_tokens().filter(|child| {
-            let child_range = child.text_range();
-            !child_range.is_empty()
-                && (child_range.start() <= offset && offset <= child_range.end())
-        });
-
-        let left = children.next().unwrap();
-        let right = children.next();
-        assert!(children.next().is_none());
-
-        if let Some(right) = right {
-            match (left.token_at_offset(offset), right.token_at_offset(offset)) {
-                (TokenAtOffset::Single(left), TokenAtOffset::Single(right)) => {
-                    TokenAtOffset::Between(left, right)
+        let mut container_element = self.clone();
+        'outer: loop {
+            for child in container_element.children_with_tokens() {
+                match child {
+                    SyntaxElement::Token(token) => {
+                        let range: TextRange = token.text_range();
+                        if range.contains_inclusive(offset) {
+                            return if offset == range.start() {
+                                match token.prev_token() {
+                                    Some(prev_token) => TokenAtOffset::Between(prev_token, token),
+                                    None => TokenAtOffset::Single(token),
+                                }
+                            } else if offset == range.end() {
+                                match token.next_token() {
+                                    Some(next_token) => TokenAtOffset::Between(token, next_token),
+                                    None => TokenAtOffset::Single(token),
+                                }
+                            } else {
+                                TokenAtOffset::Single(token)
+                            };
+                        }
+                    }
+                    SyntaxElement::Node(node) => {
+                        if node.text_range().contains_inclusive(offset) {
+                            container_element = node;
+                            continue 'outer;
+                        }
+                    }
                 }
-                _ => unreachable!(),
             }
-        } else {
-            left.token_at_offset(offset)
+            unreachable!()
```

Comment:
This should theoretically be able to be reached if there's a node with zero children... Right?

Reply:
for nodes without children that shouldn't be reached because of https://github.com/rust-analyzer/rowan/blob/be215ee19c8818e854164580f34ad3ef0d78e75a/src/cursor.rs#L452-L454

Reply:
I did find another issue related to empty nodes. I fixed it and added a couple of tests. https://github.com/rust-analyzer/rowan/pull/51/commits/2bfdd9ec4cf2c7882c6ba7ffb65425aa9c978fab